### PR TITLE
WI #2403 Move conversion of TextDocumentContentChangeEvents into TextChanges code to parser

### LINKED ISSUE #2403: Incremental tooling: merge TextDocumentContentChangeEvents and TextChangedEvent
The LSP `OnDidChangeTextDocument` method converts array of `TextDocumentContentChangeEvent` into a single instance of `TextChangedEvent` using current state of compilation document.

To allow unit testing of the whole incremental change processing, this conversion code should happen inside TypeCobol.dll so LanguageServer would only convert notification objects into semantically equivalent (or close) parser objects.

- either create a new `TextChangeType` named `RangeUpdate` to support this new type of update
- if this new type is actually not required, replace `TextChange` with a new class that matches the definition of `TextDocumentContentChangeEvent`.
  - `OnDidChangeTextDocument` will simply convert each `TextDocumentContentChangeEvent` into this new type
  - The actual processing of events that currently takes place in `OnDidChangeTextDocument` will move into `CompilationDocument.UpdateTextLines`

### PR DESCRIPTION
**Pull request by fm-117**
Fixes #2403

Note: new tests from #2404 and #2405 will have to be adapted after merge.

### CODE REVIEW

File: TypeCobol/Compiler/CompilationDocument.cs [hunk 1]
```diff
@@ -284,118 +284,115 @@ public void UpdateTextLines(TextChangedEvent textChangedEvent)
...
+                    string endOfLastLine = null;
+                    if (update.ColumnEnd < originalLastLineText.Length)
+                    {
+                        endOfLastLine = originalLastLineText.Substring(update.ColumnEnd);
+                    }
 
-                                if (lineToUpdate.CodeElements != null)
-                                    encounteredCodeElement = true; 
-                            }
+                    // Remove all the old lines
```
Comment (smedilol) [inline TypeCobol/Compiler/CompilationDocument.cs:365]: I think we should try to merge the removal and insertion of lines in order to use `TextChangeType.LineUpdated`.
The goal is to optimize by limiting the update of LineIndex of all lines after the insertion of removal.

**Review by smedilol — CHANGES_REQUESTED** [review]

File: TypeCobol.Test/Parser/Incremental/TextLineChanges/DeleteMultipleLines.inc [hunk 2]
```diff
@@ -1,9 +1,2 @@
...
-LineRemoved	5	
-LineRemoved	5	
-LineRemoved	5	
-LineRemoved	5	
-LineRemoved	5	
-LineRemoved	5	
-LineRemoved	5	
-LineInserted	5	       01 W-TechCTX TYPE TechnicalContext.
\ No newline at end of file
+(8, 26) -> (14, 26)
```
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol.Test/Parser/Incremental/TextLineChanges/DeleteMultipleLines.inc:2]: I don't understand. For this edit the client will send on single RangeUpdate that corresponds to the area selected by user to remove. Do you mean creating a test with multiple RangeUpdate like one for each line ? Or the remark is for another test ?

File: TypeCobol/Compiler/CompilationDocument.cs (hunk 1 above)
Comment (smedilol) [reply TypeCobol/Compiler/CompilationDocument.cs:365]: I've created issue #2416 for that.

File: TypeCobol.Test/Parser/Incremental/TextLineChanges/DeleteMultipleLines.inc (hunk 2 above)
[the remark this answers is not in the document]
Comment (fm-117) [reply TypeCobol.Test/Parser/Incremental/TextLineChanges/DeleteMultipleLines.inc:2]: Added the test corresponding to multiple line deletion, each one on a single RangeUpdate

**Review by smedilol — APPROVED** [review]

**Pull request merged**